source/linux/cpu_usage_sampler.c

```diff
@@ -0,0 +1,196 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/cpu_usage_sampler.h>
+
+#include <inttypes.h>
+#include <sys/sysinfo.h>
+#include <sys/types.h>
+
+/*********************************************************************************************************************
+ * Base implementation
+ ********************************************************************************************************************/
+
+struct aws_cpu_sampler_vtable {
+    int (*aws_get_cpu_sample_fn)(struct aws_cpu_sampler *sampler, double *output);
+    void (*aws_cpu_sampler_destroy)(struct aws_cpu_sampler *sampler);
+};
+
+struct aws_cpu_sampler {
+    const struct aws_cpu_sampler_vtable *vtable;
+    struct aws_allocator *allocator;
+    void *impl;
+};
+
+/*********************************************************************************************************************
+ * Linux Specific
+ ********************************************************************************************************************/
+
+struct aws_cpu_sampler_linux {
+    struct aws_cpu_sampler base;
+
+    uint64_t cpu_last_total_user;
+    uint64_t cpu_last_total_user_low;
+    uint64_t cpu_last_total_system;
+    uint64_t cpu_last_total_idle;
+    // CPU usage is reported in deltas.
+    // For the first report we have to process and cache but not send.
+    bool cpu_only_gather;
+};
+
+static void s_get_cpu_usage_linux(
+    uint64_t *total_user,
+    uint64_t *total_user_low,
+    uint64_t *total_system,
+    uint64_t *total_idle) {
+
+    FILE *file;
+    int matched_results;
+    file = fopen("/proc/stat", "r");
+    matched_results = fscanf(
+        file,
+        "cpu %" PRIu64 " %" PRIu64 " %" PRIu64 " %" PRIu64 "",
+        (uint64_t *)total_user,
+        (uint64_t *)total_user_low,
+        (uint64_t *)total_system,
+        (uint64_t *)total_idle);
+    fclose(file);
+    if (matched_results == EOF || matched_results != 4) {
+        aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+}
+
+static int aws_get_cpu_sample_fn_linux_get_uint64_delta(uint64_t *first, uint64_t *second, uint64_t *output) {
+    int64_t ret_val = AWS_OP_SUCCESS;
+    if (*first > *second) {
+        ret_val = aws_sub_u64_checked(*first, *second, output);
+    } else {
+        ret_val = aws_add_u64_checked((UINT64_MAX - *second), *first, output);
+    }
+    return ret_val;
```

Comment:
In fact, the whole function can't fail now, so let's ditch the return value and then remove all of those if-statements in the caller.